### PR TITLE
Android 13+ custom notification actions

### LINKED ISSUE #9764: [Android 13+] Missing notification player actions
### Checklist

- [X] I am able to reproduce the bug with the [latest version](https://github.com/TeamNewPipe/NewPipe/releases/latest).
- [X] I made sure that there are *no existing issues* - [open](https://github.com/TeamNewPipe/NewPipe/issues) or [closed](https://github.com/TeamNewPipe/NewPipe/issues?q=is%3Aissue+is%3Aclosed) - which I could contribute my information to.
- [X] I have read the [FAQ](https://newpipe.net/FAQ/) and my problem isn't listed.
- [X] I have taken the time to fill in all the required details. I understand that the bug report will be dismissed otherwise.
- [X] This issue contains only one bug.
- [X] I have read and understood the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md).

### Affected version

0.25.0

### Steps to reproduce the bug

Play any song,
Swipe down to see the notification player
Swipe the notification player, or just tap on it

### Expected behavior

Next/previous track buttons, pause/play, repeat, close buttons are visible.

### Actual behavior

Only next/previous track and play/pause buttons are visible. Changing it in preferences does nothing. Not mentioning that I have enough space to have all five of them.

### Screenshots/Screen recordings

photo_2023-02-09_15-26-51

### Logs

_No response_

### Affected Android/Custom ROM version

Android 13

### Affected device model

Galaxy M12

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Stypox**
#### What is it?
- [x] Bugfix (user facing)
- [ ] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
- Supersedes https://github.com/TeamNewPipe/NewPipe/pull/10567 because the code there was too hacky.
- Create class `NotificationActionData` and move there (from `NotificationUtil`) the function that turns actions from the enum format (i.e. the number stored into preferences) into the triple (action string, name, icon). This allows deduplication of such function, instead of passing arrays around like in https://github.com/TeamNewPipe/NewPipe/pull/10567.
- Set custom action providers in the media session to customize actions on Android 13+.
- Use the workaround initially suggested in https://github.com/TeamNewPipe/NewPipe/pull/10567 to basically tell the system that we're not able to handle "prev" and "next", in order to have 4 customizable action slots instead of just 2.
- On Android 13+ normal notification actions are useless, so they are not set into the notification builder anymore, to save battery
- The opposite happens on Android 12-, where media session actions are not set because they don't seem to do anything
- Exclude the play-pause and play-pause-buffering actions from ever being set to media session actions, as the play-pause-buffering action is always shown by the system. The only difference between the system-provided play-pause-buffering and ours is that when the a video completes, the replay icon is not shown, but just a normal "play".
- Currently the UI in the settings has not been changed. This can be done in another PR, which should just remove one of the 5 customizable actions and the play-pause* actions in the settings (just for Android 13+), along with a settings migration that should kick in on Android system updates.

#### Before/After Screenshots/Screen Record

<table>
  <tr>
    <th>Before</th>
    <th>After</th>
  </tr>
  <tr>
    <td>Before</td>
    <td>After</td>
  </tr>
</table>

#### Fixes the following issue(s)

- Fixes #9764

#### APK testing

The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR. You can find more info and a video demonstration [on this wiki page](https://github.com/TeamNewPipe/NewPipe/wiki/Download-APK-for-PR).

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Stypox** [comment]
> Is 5 still the max number of buttons allowed?

On Android 13+ it's 4 customizable buttons instead of 5, but at the moment the settings still show 5 actions.

> When does this happen? Upon the first time the app is opened after a system version upgrade?

Yes, but as I said it's not implemented yet and will be done in a separate PR. This way we can merge this into the RC and at least provide something that works to the user, and later cleanup the settings for Android 13+.

File: app/src/main/java/org/schabi/newpipe/player/mediasession/PlayQueueNavigator.java [hunk 1]
```diff
@@ -46,7 +47,16 @@ public PlayQueueNavigator(@NonNull final MediaSessionCompat mediaSession,
...
+        // As documented in
+        // https://developer.android.com/about/versions/13/behavior-changes-13#playback-controls
+        // starting with android 13, setting ACTION_SKIP_TO_PREVIOUS and ACTION_SKIP_TO_NEXT forces
+        // buttons 2 and 3 to be the system provided "Previous" and "Next".
+        // Thus, we pretend to not support those actions to have the ability to customize them in
+        // MediaSessionPlayerUi.updateMediaSessionActions().
+        return ACTION_SKIP_TO_QUEUE_ITEM
+                | (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU
+                        ? ACTION_SKIP_TO_NEXT | ACTION_SKIP_TO_PREVIOUS
+                        : 0);
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/player/mediasession/PlayQueueNavigator.java:59]: Is this going to hurt Bluetooth media controls like double pressing the pause button on a headset to skip to the next item?

Comment (AudricV) [reply app/src/main/java/org/schabi/newpipe/player/mediasession/PlayQueueNavigator.java:59]: Yes, this change prevents to do the action you described, I just tested the artifact built by the CI with a Bluetooth speaker.

**Comment by Stypox** [comment]
Oh, nevermind then, this approach is not really usable unfortunately. I will open a PR which just allows customizing the 2 remaining buttons.

**Pull request closed without merging**